src/NServiceBus.Core.Tests/Recoverability/FailureInfoStorageTests.cs

```diff
@@ -0,0 +1,90 @@
+namespace NServiceBus.Core.Tests.Timeout.TimeoutManager
+{
+    using System;
+    using System.Runtime.ExceptionServices;
+    using NUnit.Framework;
+
+    public class FailureInfoStorageTests
+    {
+        [Test]
+        public void When_recording_failure_initially_should_store_one_failed_attempt_and_exception()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+            var exception = new Exception();
+            var exceptionDispatchInfo = ExceptionDispatchInfo.Capture(exception);
+
+            var storage = GetFailureInfoStorage();
+
+            storage.MarkForFirstLevelRetry(messageId, exceptionDispatchInfo);
+            
+            var failureInfo = storage.GetFailureInfoForMessage(messageId);
+
+            Assert.NotNull(failureInfo);
+            Assert.AreEqual(1, failureInfo.FailedProcessingAttempts);
+            Assert.AreSame(exception, failureInfo.Exception);
+        }
+
+        [Test]
+        public void When_recording_failure_many_times_should_store_number_of_attempts_and_last_exception()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+            var secondException = new Exception();
+
+            var storage = GetFailureInfoStorage();
+
+            storage.MarkForFirstLevelRetry(messageId, ExceptionDispatchInfo.Capture(new Exception()));
+            storage.MarkForFirstLevelRetry(messageId, ExceptionDispatchInfo.Capture(secondException));
+
+            var failureInfo = storage.GetFailureInfoForMessage(messageId);
+
+            Assert.NotNull(failureInfo);
+            Assert.AreEqual(2, failureInfo.FailedProcessingAttempts);
+            Assert.AreSame(secondException, failureInfo.Exception);
+        }
+
+        [Test]
+        public void When_clearing_failure_should_return_null_on_subsequent_retrieval()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+
+            var storage = GetFailureInfoStorage();
+
+            storage.MarkForFirstLevelRetry(messageId, ExceptionDispatchInfo.Capture(new Exception()));
+            
+            var failureInfo = storage.GetFailureInfoForMessage(messageId);
+            Assert.NotNull(failureInfo);
+            
+            storage.ClearFailureInfoForMessage(messageId);
+
+            failureInfo = storage.GetFailureInfoForMessage(messageId);
+            Assert.AreSame(ProcessingFailureInfo.NullFailureInfo, failureInfo);
+        }
+
+        [Test]
+        public void When_recording_more_than_max_number_of_failures_should_remove_least_recently_used_entry()
```

Comment:
this test doesn't seem to catch the entire logic implemented in the storage? According to this test, a ring buffer would  pass this test too.